Documentation/teaching/so2/assign1-kprobe-based-tracer.rst

```diff
@@ -0,0 +1,173 @@
+==================================
+Assignment 1 - Kprobe based tracer
+==================================
+
+-  Deadline: :command:`Wednesday, 7 April 2021, 23:00`
+
+Assignment's Objectives
+=======================
+
+*  acquiring knowledge related to the instrumentation of functions in the Linux kernel (``kretprobes`` mechanism)
+*  gaining knowledge regarding the ``/proc`` file system from the Linux kernel
+*  getting more familiar with the data structures specific to the Linux kernel (``hash table``, ``list``)
+
+Statement
+=========
+
+Build a kernel operations surveillant.
+
+With this surveillant, we aim to intercept:
+
+* ``kmalloc`` and ``kfree`` calls
+* ``schedule`` calls
+* ``up`` and ``down_interruptible`` calls
+* ``mutex_lock`` and ``mutex_unlock`` calls
+
+The surveillant will hold, at the process level, the number of calls for each of the above functions.
+For the ``kmalloc`` and ``kfree`` calls the total quantity of allocated and deallocated memory will be
+shown.
+
+The surveillant will be implemented as a kernel module with the name ``tracer.ko``.
+
+Implementation details
+----------------------
+
+The interception will be done by recording a sample (``kretprobe``) for each of the above functions. The
+surveillant will retain a list/table of dispersion with the monitored processes and will account for
```

Comment:
will retain a list/hashtable

Reply:
replace with hashtable in all occurences

Reply:
Done.